include/boundary_region.hxx

```diff
@@ -20,23 +20,30 @@ enum BndryLoc {BNDRY_XIN=1,
                BNDRY_PAR_FWD=16,   // Don't include parallel boundaries
                BNDRY_PAR_BKWD=32};
 
-class BoundaryRegionBase {
+/// Describes a region of the boundary, and a means of iterating over it
+class BoundaryRegion {
 public:
-
-  BoundaryRegionBase() = delete;
-  BoundaryRegionBase(std::string name, Mesh *passmesh = nullptr)
-      : localmesh(passmesh ? passmesh : mesh), label(std::move(name)) {}
-  BoundaryRegionBase(std::string name, BndryLoc loc, Mesh *passmesh = nullptr)
+  BoundaryRegion() = delete;
+  BoundaryRegion(std::string name, BndryLoc loc, Mesh *passmesh = nullptr)
       : localmesh(passmesh ? passmesh : mesh), label(std::move(name)), location(loc) {}
+  BoundaryRegion(std::string name, int xd, int yd, Mesh *passmesh = nullptr)
+      : bx(xd), by(yd), width(2), localmesh(passmesh ? passmesh : mesh), label(std::move(name)) {}
+  virtual ~BoundaryRegion() {}
 
-  virtual ~BoundaryRegionBase() {}
+  int x,y; ///< Indices of the point in the boundary
+  int bx, by; ///< Direction of the boundary [x+dx][y+dy] is going outwards
+
+  int width; ///< Width of the boundary
```

Comment:
`const`

Reply:
We can't easily make `width` a `const` because BoundaryWidth needs to change it.

I don't see a way around it: the closest to a workaround I can think of would be for BoundaryWidth to make a new BoundaryRegion for itself with a different width, but BoundaryWidth::clone doesn't know the sub-class of BoundaryRegion it has as it's bndry member, so we'd need a factory class for BoundaryRegions...

BoundaryWidth might be useful. I can at least think of one scenario: if you want to evolve fields on the y-boundaries, you need evolving `CELL_YLOW` fields at both boundaries. That's fine at ystart because the `mesh->ystart` point is at the boundary. But at yend, you'd need to evolve the `mesh->yend+1` point, but you can't. Instead, you could use the BoundaryWidth modifier to set boundary conditions on the other fields with a width of 3, allowing you to move the boundary in one point. I admit it's a bit unlikely because if I actually wanted to do something like this, I'd probably have to implement all the sheath boundary conditions by hand, but it's enough for me to think BoundaryWidth is worth keeping.

Reply:
Would another possibility be to make the width an input to the boundary constructor? As currently we can specify the value of the Dirichlet/Neumann boundary, we could add width as an optional argument. Support for keywords in the input file would be nice...

Reply:
Ah, I didn't realise there was a `BoundaryWidth` modifier. That does complicate things a little bit.

It's possible most of the `BoundaryRegion`s could be replaced with some kind of wrapper around `Region`.

Reply:
Could be nice to do as @bendudson suggests and make the width an argument to the boundary condition input. I've now replaced most of the method bodies for Boundary*::clone with two template functions, one that gets an expression from arguments and one that ignores arguments. So now there's only one function that needs changing to add a width argument. Would be easy to add 'width' now, but it seems unlikely to be used, so more sensible to wait for possible keyword arguments from options (?) rather than adding a positional argument that we have to keep supporting.

Reply:
I've added keywords in commit 115f86750f (branch `boundary-keywords`), but no testing yet. I based it on `next`, but could rebase it and merge into this branch if it's useful.

Reply:
I've now merged #1346 and implemented a 'width' keyword. Then in the next commit made several member variables (including `BoundaryRegion::width`) `const`, and deprecated the BoundaryWidth modifier (throws an exception with a message to use the 'width' keyword).